In [4]:
import pandas as pd

In [64]:
#First one for tableau data
df = pd.read_csv(r'c:\Jalpa\Data Science UOT\Case Studies\Investment_portfolio\Cust_Invest_data.csv')

#This is for Feature Engineering and Predictive modeling
#df = pd.read_csv(r'c:\Jalpa\Data Science UOT\Case Studies\Investment_portfolio\Investment_portfolio_analysis\Data\100Cust_Invest_data.csv')

#df = pd.read_csv('Cust_Invest_data.csv')
df.head

<bound method NDFrame.head of    Customer_ID                 Asset_Types            Investment_Amounts  \
0   Customer_1         Bonds, Gold, Stocks  46693.31, 40330.82, 17836.22   
1   Customer_2  Stocks, Real Estate, Bonds  18351.85, 48591.92, 31063.11   
2   Customer_3                Gold, Crypto            46372.42, 28900.68   
3   Customer_4  Stocks, Real Estate, Bonds    8083.46, 8058.24, 43781.86   
4   Customer_5       Bonds, Stocks, Crypto  34581.79, 30435.21, 19236.64   
5   Customer_6         Bonds, Gold, Stocks  22810.36, 35226.06, 18499.87   
6   Customer_7                      Stocks                      29193.55   
7   Customer_8                 Real Estate                      44638.89   
8   Customer_9                 Real Estate                      12098.61   
9  Customer_10                 Real Estate                       32525.5   

                Current_Values                     Investment_Date  
0   35567.26, 44512.46, 223.29  2023-11-27, 2024-02-05, 2024-05-

Now, need to separate each row in df so that each asset type gets its own row instead of being combined in a single row per customer. This can be achieve by exploding the multi-asset columns. 


In [13]:
import pandas as pd

# Function to split and explode multi-asset data
def separate_assets(df):
    df_expanded = df.copy()
    
    # Split columns with multiple values into lists
    df_expanded["Asset_Types"] = df_expanded["Asset_Types"].str.split(", ")
    df_expanded["Investment_Amounts"] = df_expanded["Investment_Amounts"].str.split(", ")
    df_expanded["Current_Values"] = df_expanded["Current_Values"].str.split(", ")
    df_expanded["Investment_Date"] = df_expanded["Investment_Date"].str.split(", ")
    
    # Explode the DataFrame to create separate rows per asset type
    df_expanded = df_expanded.explode(["Asset_Types", "Investment_Amounts", "Current_Values", "Investment_Date"])

    # Convert Investment_Amounts and Current_Values to float
    df_expanded["Investment_Amounts"] = df_expanded["Investment_Amounts"].astype(float)
    df_expanded["Current_Values"] = df_expanded["Current_Values"].astype(float)
    # Convert date column
    df_expanded["Investment_Date"] = pd.to_datetime(df_expanded["Investment_Date"], errors="coerce")

    
    return df_expanded.reset_index(drop=True)

# Apply transformation
df_separated = separate_assets(df)
print(df_separated)


    Customer_ID  Asset_Types  Investment_Amounts  Current_Values  \
0    Customer_1        Bonds            46693.31       35567.260   
1    Customer_1         Gold            40330.82       44512.460   
2    Customer_1       Stocks            17836.22         223.290   
3    Customer_2       Stocks            18351.85         694.050   
4    Customer_2  Real Estate            48591.92        8144.090   
5    Customer_2        Bonds            31063.11       37129.770   
6    Customer_3         Gold            46372.42        9406.130   
7    Customer_3       Crypto            28900.68       96726.230   
8    Customer_4       Stocks             8083.46         160.800   
9    Customer_4  Real Estate             8058.24       37698.550   
10   Customer_4        Bonds            43781.86       39065.980   
11   Customer_5        Bonds            34581.79       44390.550   
12   Customer_5       Stocks            30435.21         416.525   
13   Customer_5       Crypto            19236.64

In [14]:
df_separated.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21 entries, 0 to 20
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Customer_ID         21 non-null     object        
 1   Asset_Types         21 non-null     object        
 2   Investment_Amounts  21 non-null     float64       
 3   Current_Values      21 non-null     float64       
 4   Investment_Date     21 non-null     datetime64[ns]
dtypes: datetime64[ns](1), float64(2), object(2)
memory usage: 968.0+ bytes


In [15]:
df_separated.columns

Index(['Customer_ID', 'Asset_Types', 'Investment_Amounts', 'Current_Values',
       'Investment_Date'],
      dtype='object')

In [16]:
df_separated.describe()

Investment_Amounts  Current_Values                Investment_Date
count           21.000000       21.000000                             21
mean         29395.731905    28115.581190  2022-11-22 19:25:42.857142784
min           8058.240000      109.610000            2020-04-26 00:00:00
25%          18499.870000      694.050000            2021-06-18 00:00:00
50%          30435.210000    34876.990000            2023-01-15 00:00:00
75%          40330.820000    39163.760000            2024-03-19 00:00:00
max          48591.920000    96726.230000            2024-12-28 00:00:00
std          12882.554325    28817.767785                            NaN

In [17]:
df_data=df_separated

In [18]:
# export to csv
# import os
# path= os.getcwd()

df_separated.to_csv("TableauData.csv",index=False)


In [1]:
pip install tabpy

Note: you may need to restart the kernel to use updated packages.Collecting tabpy
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
   ---------------------------------------- 0.0/1.5 MB ? eta -:--:--
   ------ --------------------------------- 0.3/1.5 MB ? eta -:--:--
   --------------------------- ------------ 1.0/1.5 MB 3.1 MB/s eta 0:00:01
   ---------------------------------- ----- 1.3/1.5 MB 3.5 MB/s eta 0:00:01
   ---------------------------------- ----- 1.3/1.5 MB 3.5 MB/s eta 0:00:01
   ---------------------------------- ----- 1.3/1.5 MB 3.5 MB/s eta 0:00:01
   ---------------------------------- ----- 1.3/1.5 MB 3.5 MB/s eta 0:00:01
   ---------------------------------- ----- 1.3/1.5 MB 3.5 MB/s eta 0:00:01
   ---------------------------------------- 1.5/1.5 MB 893.8 kB/s eta 0:00:00
   ---------------------------------------- 0.0/624.3 kB ? eta -:--:--
   --------------------------------- ------ 524.3/624.3 kB 2.8 MB/s eta 0

In [22]:
#df_new= pd.read_csv()
df_new = pd.read_csv(r'c:\Jalpa\Data Science UOT\Case Studies\Investment_portfolio\TableauData.csv')


In [23]:
df_new

Customer_ID  Asset_Types  Investment_Amounts  Current_Values  \
0    Customer_1        Bonds            46693.31       35567.260   
1    Customer_1         Gold            40330.82       44512.460   
2    Customer_1       Stocks            17836.22         223.290   
3    Customer_2       Stocks            18351.85         694.050   
4    Customer_2  Real Estate            48591.92        8144.090   
5    Customer_2        Bonds            31063.11       37129.770   
6    Customer_3         Gold            46372.42        9406.130   
7    Customer_3       Crypto            28900.68       96726.230   
8    Customer_4       Stocks             8083.46         160.800   
9    Customer_4  Real Estate             8058.24       37698.550   
10   Customer_4        Bonds            43781.86       39065.980   
11   Customer_5        Bonds            34581.79       44390.550   
12   Customer_5       Stocks            30435.21         416.525   
13   Customer_5       Crypto            19236.64       96726.230   
14   Customer_6        Bonds            22810.36        5512.350   
15   Customer_6         Gold            35226.06       39163.760   
16   Customer_6       Stocks            18499.87         498.240   
17   Customer_7       Stocks            29193.55         109.610   
18   Customer_8  Real Estate            44638.89       18435.640   
19   Customer_9  Real Estate            12098.61       34876.990   
20  Customer_10  Real Estate            32525.50       40968.700   

   Investment_Date  
0       2023-11-27  
1       2024-02-05  
2       2024-05-23  
3       2023-01-15  
4       2020-05-04  
5       2024-03-11  
6       2020-04-26  
7       2022-11-06  
8       2022-04-06  
9       2020-09-02  
10      2024-12-28  
11      2021-06-18  
12      2021-02-22  
13      2023-10-03  
14      2021-10-09  
15      2024-03-19  
16      2020-05-28  
17      2022-11-02  
18      2024-12-05  
19      2024-09-03  
20      2024-07-10

In [25]:
df_new["Investment_Date"] = pd.to_datetime(df_new["Investment_Date"], errors="coerce")


# Feature Engineering
We can create new features such as:

ROI (%) = (Current_Values - Investment_Amounts) / Investment_Amounts * 100
Time-based features (e.g., Year, Month, Quarter)


In [27]:
# Calculate ROI
df_new["ROI (%)"] = (df_new["Current_Values"] - df_new["Investment_Amounts"]) / df_new["Investment_Amounts"] * 100

# Extract time-based features
df_new["Year"] = df_new["Investment_Date"].dt.year
df_new["Month"] = df_new["Investment_Date"].dt.month
df_new["Quarter"] = df_new["Investment_Date"].dt.quarter

print(df_new.head())

  Customer_ID  Asset_Types  Investment_Amounts  Current_Values  \
0  Customer_1        Bonds            46693.31        35567.26   
1  Customer_1         Gold            40330.82        44512.46   
2  Customer_1       Stocks            17836.22          223.29   
3  Customer_2       Stocks            18351.85          694.05   
4  Customer_2  Real Estate            48591.92         8144.09   

  Investment_Date    ROI (%)  Year  Month  Quarter  
0      2023-11-27 -23.827932  2023     11        4  
1      2024-02-05  10.368349  2024      2        1  
2      2024-05-23 -98.748109  2024      5        2  
3      2023-01-15 -96.218092  2023      1        1  
4      2020-05-04 -83.239827  2020      5        2  


the Prediction Goal 

Future Investment Value (Regression Problem)
Whether an investment will be Profitable or Not (Classification Problem)

Let's proceed with predicting Future Investment Value based on past data.

Train a Machine Learning Model

# Linear Regression model
Linear Regression for predicting Current_Values based on investment amount, asset type, and time-based features.

Evaluation Metrics for Regression
Since we trained a Linear Regression model to predict Current_Values, the key metrics to evaluate it are:

Mean Absolute Error (MAE): Measures the average absolute errors: Average absolute difference between actual and predicted values. Lower is better.

Mean Squared Error (MSE): Measures the average squared errors: Similar to MAE but penalizes large errors more heavily. Lower is better.

R² Score (Coefficient of Determination): Measures how well the model explains variance in the target variable. Closer to 1 is better fit.

If R² < 0.5, consider:
    - Adding more relevant features
    - Feature engineering (e.g., log transformation, polynomial features)
    - Tuning hyperparameters (e.g., learning rate, max_depth)



In [35]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error
from sklearn.metrics import r2_score
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score



In [56]:

# Encode categorical features
label_encoder = LabelEncoder()
df_new["Asset_Types"] = label_encoder.fit_transform(df_new["Asset_Types"])  # Convert asset type to numbers

# Define features (X) and target (y)
X = df_new[["Investment_Amounts", "Asset_Types", "Year", "Month", "Quarter"]]
y = df_new["Current_Values"]

# Split data into training and testing sets (80-20 split)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train a regression model
LR_model = LinearRegression()
LR_model.fit(X_train, y_train)



# Make predictions
y_pred = LR_model.predict(X_test)

# Evaluate the model
mae_LR = mean_absolute_error(y_test, y_pred)
mse_LR = mean_squared_error(y_test, y_pred)
r2_LR = r2_score(y_test, y_pred)

print(f"MAE_LR: {mae_LR}, MSE_LR: {mse_LR}")
print(f"R²_LR Score: {r2_LR:.2f}")  # Closer to 1 means better fit


# Display predicted vs actual values
LR_results = pd.DataFrame({"Actual": y_test, "Predicted": y_pred})
print(LR_results.head())





MAE_LR: 23943.354628825597, MSE_LR: 599434165.9411061
R²_LR Score: -0.56
      Actual     Predicted
0   35567.26  53510.007902
17    109.61  27115.273267
15  39163.76  14934.566004
1   44512.46  12848.926244
8     160.80  19036.434223


Interpretation:

R² = -0.56. R² measures the proportion of the variance in the dependent variable that is predictable from the independent variables. An R² score ranges from 0 to 1 (for a good fit), and negative values indicate that your model is worse than a simple mean-based model. In this case, an R² of -0.56 suggests that the model is doing poorly and is less accurate than just predicting the average of the target variable. It indicates a significant misfit and implies that your model is not capturing the underlying trends in the data.

The negative R² value suggests the model’s predictions are worse than predicting the average of the target variable, indicating that the chosen features may not be suitable, or the model is not well-specified.


MAE = 23943 means, on average, the model's predictions are off by $23943.

MSE = 599434165 is large due to squared differences but helps compare models.this suggests that there may be some significant outliers or large deviations in your model’s predictions, given the high number

This model is not performing well overall, as shown by the high MAE and MSE.



Future Predictions
Let's say we want to predict the future investment value for:

Investment Amount = $20,000
Asset Type = "Stocks"
Investment Date = March 2025


In [60]:
# # Sample data to fit the encoder
# asset_types = ["Stocks", "Bonds", "Real Estate"]  # Your list of asset types


# # Create the LabelEncoder and fit it
# label_encoder = LabelEncoder()
# label_encoder.fit(asset_types)

# Encode asset type
asset_type_encoded = label_encoder.fit_transform(["Stocks"])[0]

# Define input for prediction
future_data = pd.DataFrame([[20000, asset_type_encoded, 2025, 3, 1]], 
                           columns=["Investment_Amounts", "Asset_Types", "Year", "Month", "Quarter"])

# Predict future value
future_value = LR_model.predict(future_data)
print(f"Predicted Future Investment Value: ${future_value[0]:,.2f}")


Predicted Future Investment Value: $41,610.29


# Classification Model - Random Forest Classifier
If we want to predict whether an investment will be profitable or not (i.e., classification problem), we need different metrics:

Accuracy = (Correct Predictions) / (Total Predictions): Overall percentage of correct predictions. Works well if classes are balanced.

Precision = TP / (TP + FP) → How many predicted profitable investments were truly profitable?: Out of all predicted positives, how many were actually positive? Important if false positives are costly.

Recall (Sensitivity) = TP / (TP + FN) → How well did we find all profitable investments?: Out of all actual positives, how many did we correctly identify? Important when missing positives is risky.

F1-score = Harmonic mean of Precision and Recall. Best for imbalanced datasets.

AUC-ROC Score = Measures model discrimination between profitable vs. non-profitable investments.:Measures how well the model distinguishes between classes (1 vs. 0). Closer to 1 is better.


If Recall is low, improve it by:
    - Increasing max_depth in XGBoost
    - Adjusting class weights (e.g., scale_pos_weight)
    - Using SMOTE (Synthetic Minority Over-sampling Technique) for imbalanced datasets


Convert Regression to Classification
Let's classify investments into Profitable (1) or Not Profitable (0) based on ROI (%).


In [43]:
import numpy as np
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score
from sklearn.ensemble import RandomForestClassifier
#import xgboost as xgb

# Create a new binary target: 1 if ROI > 0, else 0
df_new["Profitable"] = np.where(df_new["ROI (%)"] > 0, 1, 0)

# Define features (X) and target (y)
X_RF_class = df_new[["Investment_Amounts", "Asset_Types", "Year", "Month", "Quarter"]]
y_RF_class = df_new["Profitable"]

# Split data
X_train_RF_class, X_test_RF_class, y_train_RF_class, y_test_RF_class = train_test_split(X_RF_class, y_RF_class, test_size=0.2, random_state=42)

# Train a classification model
clf_RF = RandomForestClassifier(n_estimators=100, random_state=42)
clf_RF.fit(X_train_RF_class, y_train_RF_class)

# Predictions
y_pred_RF_class = clf_RF.predict(X_test_RF_class)
y_pred_RF_prob = clf_RF.predict_proba(X_test_RF_class)[:, 1]  # Probability scores for ROC AUC

# Compute classification metrics
accuracy_RF = accuracy_score(y_test_RF_class, y_pred_RF_class)
precision_RF = precision_score(y_test_RF_class, y_pred_RF_class)
recall_RF = recall_score(y_test_RF_class, y_pred_RF_class)
f1_RF = f1_score(y_test_RF_class, y_pred_RF_class)
roc_auc_RF = roc_auc_score(y_test_RF_class, y_pred_RF_prob)

# Print results
print(f"Accuracy_RF: {accuracy_RF:.2f}")
print(f"Precision_RF: {precision_RF:.2f}")
print(f"Recall_RF: {recall_RF:.2f}")
print(f"F1 Score_RF: {f1_RF:.2f}")
print(f"AUC-ROC_RF: {roc_auc_RF:.2f}")


Accuracy_RF: 0.60
Precision_RF: 0.50
Recall_RF: 0.50
F1 Score_RF: 0.50
AUC-ROC_RF: 0.67


Interpretation:

Accuracy = 60% → The model classifies 60% of investments correctly.
Precision = 50% → 50% of predicted profitable investments were truly profitable.
Recall = 50% → The model correctly identifies 50% of actual profitable investments.
F1 Score = 50% → medium balance between precision and recall.
AUC-ROC = 0.67 → The model has fair discrimination ability.

XGBoost is a powerful and efficient model for both regression and classification tasks.
Need to install: pip install xgboost
# XGBoost Regression Model


In [ ]:
pip install xgboost


In [39]:
import xgboost as xgb


In [44]:

# Define features (X) and target (y)
X_xgb_reg = df_new.drop(columns=["Customer_ID", "Investment_Date", "Current_Values"])
y_xgb_reg = df_new["Current_Values"]

# Split data into train and test sets
X_train_xgb_reg, X_test_xgb_reg, y_train_xgb_reg, y_test_xgb_reg = train_test_split(X_xgb_reg, y_xgb_reg, test_size=0.2, random_state=42)

# Train XGBoost Regressor
xgb_regressor = xgb.XGBRegressor(n_estimators=100, learning_rate=0.1, max_depth=5, random_state=42)
xgb_regressor.fit(X_train_xgb_reg, y_train_xgb_reg)

# Predictions
y_pred_xgb_reg = xgb_regressor.predict(X_test_xgb_reg)

# Regression Metrics
mae_xgb_reg = mean_absolute_error(y_test_xgb_reg, y_pred_xgb_reg)
mse_xgb_reg = mean_squared_error(y_test_xgb_reg, y_pred_xgb_reg)
r2_xgb_reg = r2_score(y_test_xgb_reg, y_pred_xgb_reg)

print(f"MAE_XGB: {mae_xgb_reg:.2f}")
print(f"MSE_XGB: {mse_xgb_reg:.2f}")
print(f"R²_XGB Score: {r2_xgb_reg:.2f}")




MAE_XGB: 4574.65
MSE_XGB: 61766400.33
R²_XGB Score: 0.84


Interpretation:

R² = 0.84 means the model explains 84% of the variance in Current_Values, which is good.
MAE = 4574.65 means, on average, the model's predictions are off by $4574.65.
MSE = 61766400.33 is large due to squared differences but helps compare models.

# XGBoost Classification Modle (Predicting Profitable Investments)
If we want to predict whether an investment is Profitable (1) or Not (0):


In [62]:
# Create binary target: 1 if ROI > 0, else 0
df_new["Profitable"] = np.where(df_new["Current_Values"] > df_new["Investment_Amounts"], 1, 0)

# Define features (X) and target (y)
X_xgb_class = df_new.drop(columns=["Customer_ID", "Investment_Date", "Current_Values", "Profitable"])
y_xgb_class = df_new["Profitable"]

# Split data
X_train_xgb_class, X_test_xgb_class, y_train_xgb_class, y_test_xgb_class = train_test_split(X_xgb_class, y_xgb_class, test_size=0.2, random_state=42)

# Train XGBoost Classifier
#xgb_classifier = xgb.XGBClassifier(n_estimators=100, learning_rate=0.1, max_depth=5, random_state=42)
xgb_classifier = xgb.XGBClassifier(n_estimators=100, learning_rate=0.1, max_depth=50, random_state=42)

xgb_classifier.fit(X_train_xgb_class, y_train_xgb_class)

# Predictions
y_pred_xgb_class = xgb_classifier.predict(X_test_xgb_class)
y_pred_xgb_prob = xgb_classifier.predict_proba(X_test_xgb_class)[:, 1]  # Probability scores for ROC AUC

# Classification Metrics
accuracy_xgb = accuracy_score(y_test_xgb_class, y_pred_xgb_class)
precision_xgb = precision_score(y_test_xgb_class, y_pred_xgb_class)
recall_xgb = recall_score(y_test_xgb_class, y_pred_xgb_class)
f1_xgb = f1_score(y_test_xgb_class, y_pred_xgb_class)
roc_auc_xgb = roc_auc_score(y_test_xgb_class, y_pred_xgb_prob)

print(f"Accuracy_xgb: {accuracy_xgb:.2f}")
print(f"Precision_xgb: {precision_xgb:.2f}")
print(f"Recall_xgb: {recall_xgb:.2f}")
print(f"F1_xgb Score: {f1_xgb:.2f}")
print(f"AUC-ROC_xgb: {roc_auc_xgb:.2f}")


Accuracy_xgb: 0.60
Precision_xgb: 0.00
Recall_xgb: 0.00
F1_xgb Score: 0.00
AUC-ROC_xgb: 0.50


c:\Users\JalpaZenisha\miniconda3\envs\dsi_participant\lib\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


Accuracy: 0.60
This means the model is correctly predicting 60% of the cases. While this is decent for a baseline.

Precision: 0.00
Precision measures the proportion of positive predictions that were actually correct. A precision of 0.00 means that when the model predicts a positive outcome, it is always wrong. This suggests that the model might not be predicting positive outcomes at all, or it could be severely imbalanced in its predictions.

Recall: 0.00
Recall measures the proportion of actual positive cases that the model correctly identifies. A recall of 0.00 indicates that the model is not identifying any of the true positive cases.

F1 Score: 0.00
The F1 score is the harmonic mean of precision and recall. Since both precision and recall are 0.00, the F1 score is also 0.00, meaning the model is performing very poorly in terms of its positive class predictions.

AUC-ROC: 0.50
AUC-ROC is a measure of the model's ability to distinguish between positive and negative classes. A value of 0.50 means the model is performing no better than random chance in distinguishing between the classes.



Possible Reasons and Solutions:

Class Imbalance: If you are working with imbalanced classes (for example, many more negatives than positives), the model may predict only the majority class. You could address this by:

Using techniques like SMOTE (Synthetic Minority Over-sampling Technique) or undersampling the majority class.
Adjusting class weights in the XGBoost model to penalize misclassifying the minority class more heavily.
Model Configuration or Hyperparameters:

The hyperparameters might need adjustment. For example, you may want to tune parameters like max_depth, learning_rate, n_estimators, and scale_pos_weight to improve performance.
Data Quality or Feature Engineering:

Ensure that your features are well-processed and relevant to the prediction task. You might need to improve feature engineering, perform feature selection, or handle missing data more effectively.
Threshold Adjustment:

The default threshold for classification might not be optimal. You can try adjusting the threshold for classification, which might help balance precision and recall.
Model Evaluation:

Check the confusion matrix to better understand where the model is making errors. This can provide insight into the specific issues (e.g., misclassifying all positives as negatives).


# Hyperparameter Tuning (Using Grid Search)
To further optimize the XGBoost model, we can perform Grid Search to find the best parameters.


In [49]:
from sklearn.model_selection import GridSearchCV

# Define parameter grid
param_grid = {
    'max_depth': [3, 5, 7],
    'learning_rate': [0.01, 0.1, 0.2],
    'n_estimators': [50, 100, 200]
}

# Perform Grid Search for Classification
grid_search = GridSearchCV(xgb.XGBClassifier(random_state=42), param_grid, cv=3, scoring="accuracy", verbose=1, n_jobs=-1)
grid_search.fit(X_train_xgb_class, y_train_xgb_class)

# Best parameters
print("Best Parameters:", grid_search.best_params_)

# Train model with best parameters
best_xgb_class = grid_search.best_estimator_
y_pred_best_xgb_class = best_xgb_class.predict(X_test_xgb_class)

# Evaluate best model
accuracy_best = accuracy_score(y_test_xgb_class, y_pred_best_xgb_class)
print(f"Tuned Model Accuracy: {accuracy_best:.2f}")


Fitting 3 folds for each of 27 candidates, totalling 81 fits
Best Parameters: {'learning_rate': 0.01, 'max_depth': 3, 'n_estimators': 100}
Tuned Model Accuracy: 0.60


# Model Deployment
Now that the model is trained, we need to save it and deploy it for real-world use.

1️. Save and Load the Model
To use the model later, save it as a .json or .pkl file.

In [ ]:
import pickle

# Save the model
with open("xgboost_model.pkl", "wb") as file:
    pickle.dump(xgb_classifier, file)

# Load the model
with open("xgboost_model.pkl", "rb") as file:
    loaded_model = pickle.load(file)

# Use the loaded model for predictions
y_pred_loaded = loaded_model.predict(X_test_class)


# 2️. Deploy with a Simple Flask API

create an API to serve predictions using Flask.




In [ ]:
pip install flask


In [ ]:
#  Create a app.py file:

from flask import Flask, request, jsonify
import pickle
import pandas as pd

# Load the trained model
with open("xgboost_model.pkl", "rb") as file:
    model = pickle.load(file)

app = Flask(__name__)

@app.route('/predict', methods=['POST'])
def predict():
    data = request.get_json()  # Expecting JSON input
    df = pd.DataFrame(data)  # Convert to DataFrame
    
    prediction = model.predict(df)
    
    return jsonify({'prediction': prediction.tolist()})

if __name__ == '__main__':
    app.run(debug=True)


In [ ]:
# 📌 Run the API:

python app.py

# This will create an API endpoint (/predict) that takes JSON input and returns predictions.



# 3. Deploy Using FastAPI (Optional)
If you need a faster API:


In [ ]:
pip install fastapi uvicorn


Create main.py:

In [ ]:
from fastapi import FastAPI
import pickle
import pandas as pd

app = FastAPI()

with open("xgboost_model.pkl", "rb") as file:
    model = pickle.load(file)

@app.post("/predict")
async def predict(data: dict):
    df = pd.DataFrame([data])
    prediction = model.predict(df)
    return {"prediction": prediction.tolist()}



Run with:

In [ ]:
uvicorn main:app --reload

Then test with:

In [ ]:
curl -X 'POST' 'http://127.0.0.1:8000/predict' -H 'Content-Type: application/json' -d '{"Investment_Amounts": 5000, "Year": 2024, "Month": 2, "Quarter": 1}'


# Next Steps
1️⃣ Optimize Feature Engineering (Try different transformations like scaling, log transformations, or PCA)
2️⃣ Deploy on Cloud (Use AWS Lambda, Google Cloud Run, or Azure Functions)
3️⃣ Automate Model Retraining (Use a scheduled job to update predictions)
4️⃣ Integrate with Tableau (Expose API for real-time updates in Tableau dashboards)

# Deploying XGBoost Model on the Cloud (AWS, GCP, or Azure)
To integrate XGBoost model with Tableau, we need to deploy it on the cloud so that Tableau can access predictions via an API. Below are the steps for deploying on AWS, but the process is similar for GCP or Azure.

**Step 1: Choose a Cloud Service**

- AWS Lambda + API Gateway (Serverless, cost-effective)
- AWS EC2 (More control, but requires manual setup)
- Google Cloud Run (Scalable serverless option)
- Azure Functions (Similar to AWS Lambda)

For simplicity, let’s deploy on AWS using Flask and EC2. 

**Step 2: Prepare Your XGBoost Model**
Convert your trained XGBoost model into a deployable format.


In [ ]:
import pickle

# Save the trained XGBoost model
with open("xgboost_model.pkl", "wb") as file:
    pickle.dump(xgb_classifier, file)


**Step 3: Set Up an AWS EC2 Instance**

1️  Log in to AWS Console

2️  Go to EC2 Dashboard → Click Launch Instance

3️  Choose Ubuntu 22.04 as OS

4️  Select t2.micro (Free-tier eligible)

5️  Configure Security Group:

    - Allow HTTP (port 80) and Custom TCP (port 5000)
    - Allow SSH (port 22) for remote access

6️  Click Launch, then Connect to Instance via SSH:




In [ ]:
ssh -i "your-key.pem" ubuntu@your-ec2-public-ip


**Step 4: Install Required Libraries**

Once inside your EC2 instance, install dependencies:


In [ ]:
sudo apt update && sudo apt upgrade -y
sudo apt install python3-pip -y
pip3 install flask pandas xgboost pickle-mixin


**Step 5: Deploy Flask API**

Create a new file app.py and paste this code:


In [ ]:
from flask import Flask, request, jsonify
import pickle
import pandas as pd

# Load the trained model
with open("xgboost_model.pkl", "rb") as file:
    model = pickle.load(file)

app = Flask(__name__)

@app.route('/predict', methods=['POST'])
def predict():
    data = request.get_json()  # Expecting JSON input
    df = pd.DataFrame(data)  # Convert to DataFrame
    prediction = model.predict(df)
    return jsonify({'prediction': prediction.tolist()})

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=5000, debug=True)


**Step 6: Start API Server**
Run the Flask app:



In [ ]:
python3 app.py


Your API is now running on  http://your-ec2-public-ip:5000/predict.

**Step 7: Test the API**

On your local machine, send a request using Postman or curl:


In [ ]:
curl -X POST "http://your-ec2-public-ip:5000/predict" \
     -H "Content-Type: application/json" \
     -d '[{"Investment_Amount": 5000, "Year": 2024, "Month": 2, "Quarter": 1}]'


Expected Response:

{"prediction": [1]}

This means Investment_Amount = $5000 is predicted to be Profitable (1).



**Step 8: Connect API to Tableau**
Now, integrate the API into Tableau.

In Tableau
- Open Tableau Desktop
- Go to Data Source → Select "Web Data Connector"
- Use a simple Python Flask API connector or a Tableau Web Data Connector (WDC)
- Fetch predictions dynamically and display in a Tableau Dashboard


Connecting Your XGBoost API to Tableau via Web Data Connector (WDC)
To display your XGBoost model’s predictions in Tableau, we need to create a Web Data Connector (WDC) that fetches data from the deployed Flask API on AWS.

🔹 Step 1: Install Tableau Web Data Connector SDK
Since Tableau does not natively support REST APIs, we need a WDC (Web Data Connector) to interact with the Flask API.

1️⃣ Install Tableau WDC SDK on your local machine:


In [ ]:
npm install -g tableauwdc


In [ ]:
#2️⃣ Clone the Web Data Connector SDK:

git clone https://github.com/tableau/webdataconnector.git
cd webdataconnector/Examples



**Step 2: Create a New Web Data Connector (WDC)**
Inside the webdataconnector/Examples folder, create a new file flask_wdc.html.

flask_wdc.html


In [ ]:
<!DOCTYPE html>
<html>
<head>
    <title>XGBoost Predictions WDC</title>
    <script src="https://connectors.tableau.com/libs/tableauwdc-2.3.latest.js"></script>
    <script>
        (function() {
            var myConnector = tableau.makeConnector();

            myConnector.getSchema = function(schemaCallback) {
                var cols = [
                    { id: "Investment_Amount", alias: "Investment Amount ($)", dataType: tableau.dataTypeEnum.float },
                    { id: "Year", alias: "Year", dataType: tableau.dataTypeEnum.int },
                    { id: "Month", alias: "Month", dataType: tableau.dataTypeEnum.int },
                    { id: "Quarter", alias: "Quarter", dataType: tableau.dataTypeEnum.int },
                    { id: "Prediction", alias: "Predicted Outcome", dataType: tableau.dataTypeEnum.string }
                ];

                var tableSchema = {
                    id: "investmentPredictions",
                    alias: "Predictions from XGBoost Model",
                    columns: cols
                };

                schemaCallback([tableSchema]);
            };

            myConnector.getData = function(table, doneCallback) {
                var apiUrl = "http://your-ec2-public-ip:5000/predict";

                fetch(apiUrl, {
                    method: "POST",
                    headers: {
                        "Content-Type": "application/json"
                    },
                    body: JSON.stringify([
                        {"Investment_Amount": 10000, "Year": 2024, "Month": 2, "Quarter": 1},
                        {"Investment_Amount": 20000, "Year": 2024, "Month": 3, "Quarter": 1}
                    ])
                })
                .then(response => response.json())
                .then(data => {
                    var tableData = [];
                    for (var i = 0; i < data.prediction.length; i++) {
                        tableData.push({
                            "Investment_Amount": [10000, 20000][i],
                            "Year": 2024,
                            "Month": [2, 3][i],
                            "Quarter": 1,
                            "Prediction": data.prediction[i] === 1 ? "Profitable" : "Risky"
                        });
                    }
                    table.appendRows(tableData);
                    doneCallback();
                });
            };

            tableau.registerConnector(myConnector);

            document.querySelector("#submitButton").addEventListener("click", function() {
                tableau.connectionName = "XGBoost Predictions Data";
                tableau.submit();
            });
        })();
    </script>
</head>
<body>
    <h1>XGBoost Investment Predictions</h1>
    <button id="submitButton">Get Data</button>
</body>
</html>


**Step 3: Host WDC on a Local Web Server**
Since Tableau cannot directly open local .html files, we need to serve it using a simple Python HTTP server.

1️⃣ Navigate to the directory where flask_wdc.html is located:


In [ ]:
cd /path/to/webdataconnector/Examples


2️⃣ Start a local web server:

python3 -m http.server 8000

3️⃣ Your WDC is now available at:

 http://localhost:8000/flask_wdc.html



**Step 4: Connect to Tableau**

1️⃣ Open Tableau Desktop

2️⃣ Go to "Connect" → Select "Web Data Connector"

3️⃣ Enter the URL:

http://localhost:8000/flask_wdc.html

4️⃣ Click "Get Data" and wait for predictions to load

5️⃣ Click "Update Extract" and then go to Sheet 1 to create visualizations



**Step 5: Deploy WDC to AWS (Optional)**
To access the WDC remotely, deploy it to AWS S3 or an EC2 instance.

1️⃣ Copy flask_wdc.html to your AWS EC2:


In [ ]:
scp -i "your-key.pem" flask_wdc.html ubuntu@your-ec2-public-ip:/var/www/html/

# 2️⃣ Access it in Tableau using:

http://your-ec2-public-ip/flask_wdc.html


**Next Steps:**
- Fine-tune XGBoost model
- Deploy on AWS Lambda (for serverless API)
- Automate API requests with Tableau Extract Refresh
